turftopic/encoders/e5.py

```diff
@@ -30,11 +30,64 @@ class E5Encoder(SentenceTransformer, ExternalEncoder):
     """
 
     def __init__(
-        self, model_name: str = "intfloat/multilingual-e5-small", instruction: Optional[str] = None, **kwargs
+        self, model_name: str = "intfloat/multilingual-e5-small", **kwargs
+    ):
+        super().__init__(model_name, **kwargs)
+
+    def encode(self, sentences: list[str], **kwargs) -> np.ndarray:
+        """
+        Parameters
+        ----------
+        sentences: list[str]
+            Input text
+
+        Notes
+        -----
+        See docs for `SentenceTransformer.encode` for available **kwargs  
+        https://www.sbert.net/docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode
+        """
+        sentences = [
+            f"query: {sentence}" for sentence in sentences
+        ]
+        return super().encode(sentences, **kwargs)
+
+
+class E5InstructionalEncoder(SentenceTransformer):
+    """Encoder model oriented at using E5 instructional models, adds a query prefix, and an optional instruction,
+    following the format: "Instruction: {instruction} \nQuery: {document}"
+
+    ```python
+    from turftopic.encoders import E5InstructionalEncoder
+    from turftopic import GMM
+
+    model = GMM(10, encoder=E5InstructionalEncoder())
+    ```
+
+    Parameters
+    ----------
+    model_name: str
+        Embedding model to use. 
+        Either a SentenceTransformers pre-trained models or a model from HuggingFace Hub.
+
+    instruction: Optional[str]
+        Instruction to prefix the query with.
+        By default `"Instruction: "`
+
+    Notes
+    -----
+    See available E5-based sentence transformers on Hugging Face Hub:  
+    https://huggingface.co/models?library=sentence-transformers&sort=trending&search=e5
+
+    Instruction templates:  
+    https://github.com/microsoft/unilm/blob/9c0f1ff7ca53431fe47d2637dfe253643d94185b/e5/utils.py#L106
+    """
+
+    def __init__(
+        self, model_name: str = "intfloat/multilingual-e5-large-instruct", instruction: Optional[str] = None, **kwargs
     ):
         super().__init__(model_name, **kwargs)
         if instruction is not None:
-            self.instruction = f"Instruction: {instruction} "
+            self.instruction = f"Instruction: {instruction}"
```

Comment:
This feels a bit weird. You template the query in the list comprehension below, but you template the instruction up here. Also what happens if `instruction is None` how does that show up in the f-string? Also imagine this kinda scenario:

```python
encoder1 = E5InstructionalEncoder("some_model", instruction="bla bla bla bla")
encoder2 = E5InstructionalEncoder("some_other_model", instruction=encoder1.instruction)
```

And then boom, instructionception:
"Instruction: Instruction: bla bla bla"

Reply:
you're right!
Fixed. 
How about prefixing documents in a static method?

```
class E5InstructionalEncoder(SentenceTransformer):
    """Encoder model oriented at using E5 instructional models, adds a query prefix, and an optional instruction.

    ```python
    from turftopic.encoders import E5InstructionalEncoder
    from turftopic import GMM

    model = GMM(10, encoder=E5InstructionalEncoder())
    ```

    Parameters
    ----------
    model_name: str
        Embedding model to use.
        Either a SentenceTransformers pre-trained models or a model from HuggingFace Hub.

    instruction: Optional[str]
        One-sentence instruction that describes the task.
        The encoder model's performance will degrade if an instruction is not specified.
        By default `None`.

    Notes
    -----
    See available E5-based sentence transformers on Hugging Face Hub:
    https://huggingface.co/models?library=sentence-transformers&sort=trending&search=e5

    Instruction templates:
    https://github.com/microsoft/unilm/blob/9c0f1ff7ca53431fe47d2637dfe253643d94185b/e5/utils.py#L106
    """

    def __init__(
        self,
        model_name: str = "intfloat/multilingual-e5-large-instruct",
        instruction: Optional[str] = None,
        **kwargs,
    ):
        super().__init__(model_name, **kwargs)
        if instruction is not None:
            self.instruction = instruction
        else:
            self.instruction = ""

    @staticmethod
    def _prefix_document(instruction: str, query: str) -> str:
        """Structure input text with E5 prefixes

        Parameters
        ----------
        instruction : str
            one-sentence instruction that describes the task.

        query : str
            text to work on
        """
        return f"Instruct: {instruction}\nQuery: {query}"

    def encode(self, sentences: list[str], **kwargs) -> np.ndarray:
        """
        Parameters
        ----------
        sentences: list[str]
            Input text

        Notes
        -----
        See docs for `SentenceTransformer.encode` for available **kwargs
        https://www.sbert.net/docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode
        """

        sentences = [
            self._prefix_document(self.instruction, sentence)
            for sentence in sentences
        ]

        return super().encode(sentences, **kwargs)
```